hail/src/main/scala/is/hail/annotations/RegionMemory.scala

```diff
@@ -0,0 +1,210 @@
+package is.hail.annotations
+
+import is.hail.utils._
+
+final class RegionMemory(pool: RegionPool) extends AutoCloseable {
+  private val usedBlocks = new ArrayBuilder[Long](4)
+  private val bigChunks = new ArrayBuilder[Long](4)
+
+  private var totalChunkMemory = 0L
+  private var currentBlock: Long = 0L
+  private var offsetWithinBlock: Long = _
+
+  // blockThreshold and blockByteSize are mutable because RegionMemory objects are reused with different sizes
+  var blockSize: Region.Size = -1
+  private var blockThreshold: Long = _
+  private var blockByteSize: Long = _
+
+  private val references = new ArrayBuilder[RegionMemory](4)
+  private var referenceCount: Long = _
+
+  def allocateNewBlock(): Unit = {
+    if (currentBlock != 0)
+      usedBlocks += currentBlock
+    currentBlock = pool.getBlock(blockSize)
+  }
+
+  def getCurrentBlock(): Long = currentBlock
+
+  private def allocateBigChunk(size: Long): Long = {
+    val o = pool.getChunk(size)
+    bigChunks += o
+    totalChunkMemory += size
+    o
+  }
+
+  def allocate(n: Long): Long = {
+    val r = if (offsetWithinBlock + n <= blockByteSize) {
+      val o = currentBlock + offsetWithinBlock
+      offsetWithinBlock += n
+      o
+    } else {
+      if (n <= blockThreshold) {
+        allocateNewBlock()
+        offsetWithinBlock = n
+        currentBlock
+      } else
+        allocateBigChunk(n)
+    }
+    r
+  }
+
+  def allocate(a: Long, n: Long): Long = {
+    val alignedOff = (offsetWithinBlock + a - 1) & ~(a - 1);
+    val r = if (alignedOff + n <= blockByteSize) {
+      val off = currentBlock + alignedOff
+      offsetWithinBlock = alignedOff + n
+      off
+    } else {
+      if (n <= blockThreshold) {
+        allocateNewBlock()
+        offsetWithinBlock = n
+        currentBlock
+      } else {
+        allocateBigChunk(n)
+      }
+    }
+    r
+  }
+
+  private def isFreed: Boolean = blockSize == -1
+
+  def freeMemory(): Unit = {
```

Comment:
it feels kind of weird to me that this is public, since it feels like it's mostly an internal function. Actually, I think RegionMemory itself could probably be package-private, which would let Region/RegionPool access all of its stuff without exposing the memory management elsewhere.

Reply:
yeah, that's what I want.